### PR TITLE
DrivAerML Coordinate Noise Augmentation (Input Perturbation Regularization)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

DrivAerML trains on 50k surface points per mesh, where each point has xyz coordinates and surface normals. The model uses Fourier positional encoding on the coordinates. Currently, the model sees the EXACT same point coordinates every time it processes a given mesh.

By adding small Gaussian noise (σ=0.001–0.005 in normalized coordinate space) to the xyz coordinates during training, we force the model to learn surface field predictions that are locally smooth and robust to small position perturbations. This is:

1. **A form of data augmentation** — each epoch sees slightly different point positions, increasing effective data diversity
2. **Implicit regularization** — forces smooth learned representations
3. **Physics-motivated** — aerodynamic surface fields should be spatially continuous; if moving a point by 0.1mm changes the prediction dramatically, the model has overfit to exact positions
4. **Affects Fourier features** — noise on coordinates creates noisy Fourier features, preventing over-reliance on specific frequency components

At test time, use clean (unperturbed) coordinates.

## Baseline

Current best DrivAerML (PR #3072, eren/dm-ema-9995-gc05):
- `val_primary/surface_rel_l2_pct`: **3.833%** at epoch 511
- test: 4.685%
- W&B run: ncl1dh88
- Config: AdamW lr=5e-4, gc=0.5, no WD, 4L/512d/8H, EMA=0.9995, T_max=30, 394 batches/epoch

**Target to beat: 3.833%**. External target: <3.71% (AB-UPT). Gap: ~0.123 pp.

## Known Bug Fix

There is a `primary_metric_key` shadowing bug in `train.py` where a local variable (line ~1646–1648) shadows the function (line ~1428), causing an `UnboundLocalError`. Fix by renaming the local variable to `best_tracking_metric_key` if you encounter it.

## Implementation

1. Add `--coord-noise-std` flag (float, default 0.0 — no noise)
2. During training only (not eval), add Gaussian noise to xyz coordinates:
   ```python
   xyz = xyz + torch.randn_like(xyz) * noise_std
   ```
3. Apply noise **BEFORE** Fourier encoding (so the Fourier features are computed on noisy positions)
4. Do **NOT** add noise to normals or target values — only to position coordinates
5. Log `coord_noise_std` to W&B config

## Trial 1 — σ=0.001 (conservative: ~0.1% of coordinate range)

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --coord-noise-std 0.001 \
  --wandb_group dm-coord-noise
```

## Trial 2 — σ=0.005 (moderate: ~0.5% of coordinate range)

Same as Trial 1 but with `--coord-noise-std 0.005`:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --ema-decay 0.9995 \
  --coord-noise-std 0.005 \
  --wandb_group dm-coord-noise
```

## Reporting

Report `val_primary/surface_rel_l2_pct` at best epoch for both trials. Target: beat **3.833%**.

Use `--wandb_group dm-coord-noise` so both trials appear together in W&B.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Trial 1: σ=0.001 (coord_noise_std=0.001)

| Metric | Baseline | Trial 1 (best) | Delta |
|--------|----------|-----------------|-------|
| val_primary/surface_rel_l2_pct | 3.833% (ep511) | **6.450%** (ep171) | +2.617 pp (1.68x worse) |

- **W&B run:** `hapy2g9k`
- **Status:** Diverged at epoch 172. Grad norms exploded exponentially: 0.5 → 9.2 → 32.8 → 1,354 → 733K → 1.9M over 5 epochs. Killed at epoch 176.
- **Notable:** Had an earlier instability spike at epochs 124-137 (peaked at 10.68%) from which it fully recovered. The second spike at epoch 172 proved fatal.
- **Peak memory:** ~2.8 GB RSS per worker (GPU VRAM not explicitly logged)

### Trial 2: σ=0.005 (coord_noise_std=0.005)

| Metric | Baseline | Trial 2 (best) | Delta |
|--------|----------|-----------------|-------|
| val_primary/surface_rel_l2_pct | 3.833% (ep511) | **21.44%** (ep23) | +17.61 pp (5.59x worse) |

- **W&B run:** `qfu1j89b`
- **Status:** Diverged by epoch ~25. Grad norms reached Infinity by epoch ~52 and never recovered. Steady-state val ~62.1% (random outputs). Killed at epoch 170.

### Commands used

```bash
# Trial 1
CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --coord-noise-std 0.001 \
  --agent alphonse --wandb-name "alphonse/dm-coord-noise-001" --wandb-group dm-coord-noise

# Trial 2
CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 0.5 --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --ema-decay 0.9995 --coord-noise-std 0.005 \
  --agent alphonse --wandb-name "alphonse/dm-coord-noise-005" --wandb-group dm-coord-noise
```

### What happened

**This hypothesis is dead.** Both noise levels caused catastrophic training instability and eventual divergence. The coordinate noise interacts badly with the Fourier positional encoding — perturbing coordinates by even σ=0.001 changes the high-frequency Fourier features (freq=32.0) by up to 32× the noise magnitude, creating large downstream perturbations that destabilize training.

Key observations:
1. **Higher noise diverges faster:** σ=0.005 diverged by epoch 25; σ=0.001 survived to epoch 171 before diverging.
2. **Instability is intermittent then catastrophic:** Trial 1 had a recoverable spike (ep124-137) before the fatal one (ep172+). The grad norm explosion is exponential once triggered.
3. **Fourier amplification is the root cause:** The `append_fourier_features` function multiplies raw coordinates by frequencies [0.5, 2.0, 8.0, 32.0]. Noise at σ=0.001 becomes σ=0.032 at the highest frequency — a 32× amplification. At σ=0.005, this becomes σ=0.16, which is enormous relative to the normalized coordinate range.
4. **EMA provided only checkpoint protection, not stability:** EMA preserved the best checkpoint but couldn't prevent the underlying model from diverging.
5. **Grad clipping was overwhelmed:** Despite grad_clip=0.5, norms reached 733K+ and Infinity, indicating the gradient explosion was too fast for per-step clipping to contain.

### Suggested follow-ups

1. **Frequency-aware noise scaling:** If coordinate noise is revisited, scale noise inversely with the max Fourier frequency: σ_effective = σ / max_freq. This would keep Fourier feature perturbations bounded.
2. **Feature-space noise instead:** Apply noise to the hidden representations after the Fourier encoding, where the model has learned representations that may be more robust to small perturbations.
3. **Noise scheduling:** Start with zero noise and slowly increase, or use a warmup period before enabling noise, to let the model build stable representations first.
4. **Lower noise magnitude:** σ=0.0001 or σ=0.0002 might survive but the benefit at such tiny scales is questionable.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3256 because ADVISOR: Closing — coordinate noise augmentation is fundamentally incompatible with Fourier positional encoding. σ=0.001 → 6.45% best (+68%, crashed ep172), σ=0.005 → 21.44% best (+459%, 117/168 grad norms were Infinity). Root cause: Fourier frequencies up to 32.0 amplify coordinate noise by 32x — σ=0.001 in coord space becomes σ=0.032 in feature space at the highest frequency. This is structural, not tunable.

Your analysis of the Fourier amplification mechanism was spot-on. The feature-space noise idea you suggested (perturbing hidden representations AFTER Fourier encoding) is the correct reformulation — it avoids the amplification problem entirely..
